roll/parser.py

```diff
@@ -0,0 +1,234 @@
+        if len(pairs) == 1:
+            return op
+        return rec_operator(op, pairs[1:])
+
+    item = xs[0]
+    suffix = xs[1]
+    return rec_operator(item, suffix)
+
+
+def bin_operator_right(xs):
+    """xs = [item, [[sep, item], ... ]]"""
+
+    def rec_operator(pairs, right):
+        if len(pairs) == 0:
+            return right
+        op = TokenOperator(pairs[-1][0], [pairs[-1][1], right])
+        if len(pairs) == 1:
+            return op
+        return rec_operator(pairs[:-1], op)
+
+    item = xs[0]
+    suffix = xs[1]
+    separators = [pair[0] for pair in suffix]
+    items = [item] + [pair[1] for pair in suffix]
+    right = items[-1]
+    prefix = [[separators[i], items[i]] for i in range(len(separators))]
+    return rec_operator(prefix, right)
+
+
+def mon_operator(xs):
+    """xs = [unary_op, unary] OR primary"""
+    if not isinstance(xs, list):
+        return xs
+    return TokenOperator(xs[0], [xs[1]])
+
+
+def maybe_dice(xs):
+    """unary & opt("d" >> unary)
+    xs = [unary, [unary]]
+    """
+    if xs[1] == []:
+        return xs[0]
+    return TokenRoll(xs[0], xs[1][0])
+
+
+def maybe_ternary(xs):
+    """case & opt("?" >> expr << ":" & expr)
+    xs = [case, []]
+    """
+    if xs[1] == []:
+        return xs[0]
+    return TokenTernary(xs[0], xs[1][0][0], xs[1][0][1])
+
+
+def maybe_case(xs):
+    """unary & opt(":" >> "(" >> rep1sep(case_pair, ";") << ")")
+    xs = [unary, [[]]]
+    """
+    if xs[1] == []:
+        return xs[0]
+    return TokenCase(xs[0], xs[1][0])
+
+
+def let(xs):
+    """assignment = identifier << "=" & expr
+    let_stmt = "let" >> rep1sep(assignment, ";") << "in" & expr
+    xs = [[[id, expr]*], expr]
+    """
+    decls = xs[0]
+    expr = xs[1]
+    new_env = [Assignment(decl[0], decl[1]) for decl in decls]
+    return TokenLet(new_env, expr)
+
+
+def anon(xs):
+    """rep1(identifier) & expr
+    xs = [[id], expr]
+    """
+    ids = xs[0]
+    expr = xs[1]
+    if len(ids) == 1:
+        return TokenFunction(ids[0], expr)
+    return TokenFunction(ids[0], anon([ids[1:], expr]))
+
+
+def maybe_application(xs):
+    """xs = [expr, expr, ...]"""
+    if len(xs) == 1:
+        return xs[0]
+    return TokenApplication(xs[0], xs[1:])
+
+
+def func_decl_or_type(xs):  #
+    """identifier & func_decl
+    xs = [id, expr]
+    """
+    return Assignment(xs[0], xs[1])
+
+
+class ProgramParser(TextParsers):
+    split1 = lambda item, separator: item & rep(separator & item)
+    split = lambda item, separator: opt(split1(item, separator))
+
+    identifier = reg(r"[a-zA-Z]\w*")
+
+    string = reg(r'".*?(?<!\\)(\\\\)*?"') | reg(r"'.*?(?<!\\)(\\\\)*?'") > (
+        lambda s: TokenString(s[1:-1])
+    )
+
+    num_int = reg(r"\d+") > int
+    num_float = reg(r"(\d*\.\d+|\d+\.\d*)") > float
+    num_positive = num_float | num_int
+    num_negative = "-" >> num > (lambda x: -x)
+    num = num_negative | num_positive
+    number = num > TokenNumber

```

Comment:
It's worth adding a https://parsita.drhagen.com/miscellaneous_parsers/#fwd-forward-declaration here to make IDEs happier, and similar within this class.
```suggestion
    num = fwd()
    num_int = reg(r"\d+") > int
    num_float = reg(r"(\d*\.\d+|\d+\.\d*)") > float
    num_positive = num_float | num_int
    num_negative = "-" >> num > (lambda x: -x)
    num = num_negative | num_positive
    number = num > TokenNumber
```

Reply:
Doing this results in `TypeError: 'ForwardDeclaration' object is not callable`

Currently won't be fixed, but I will return to it.